cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/ClusterModel.java

```diff
@@ -714,6 +716,57 @@ public Replica createReplica(String rackId, int brokerId, TopicPartition tp, int
     return replica;
   }
 
+  /**
+   * Add an imaginary replica to the cluster.The replica does not correspond to any existing replica in the cluster, but
+   * a replica we are going to add to the cluster. The imaginary replica's original broker will not be any existing broker
+   * so that it will be treated as an immigrant replica for whatever broker it is assigned to and grant goals greatest
+   * freedom to allocate to an existing broker.
+   *
+   * @param topicPartition The topic partition of the imaginary replica.
+   * @param brokerId       Id of broker which the imaginary replica is assigned to.
+   */
+  public void addReplica(TopicPartition topicPartition, int brokerId) {
+    Replica replica = new Replica(topicPartition, IMAGINARY_BROKER, false);
+    replica.setBroker(broker(brokerId));
+    // Update partition info.
+    Partition partition = _partitionsByTopicPartition.get(topicPartition);
+    partition.addFollower(replica, partition.followers().size());
+
+    // Update replication factor statistics in cluster.
+    int replicationFactor = Math.max(_replicationFactorByTopic.get(topicPartition.topic()), partition.replicas().size());
+    _replicationFactorByTopic.put(topicPartition.topic(), replicationFactor);
+    _maxReplicationFactor = Math.max(_maxReplicationFactor, replicationFactor);
+
+    // Populate the load for imaginary replica.
+    replica.setMetricValues(partition.leader().load().loadByWindows(), partition.leader().load().windows());
+    replica.makeFollower(true);
+
+    // Add this replica and associated load to the assigned broker / assigned rack / cluster.
+    broker(brokerId).rack().addReplica(replica);
+    _load.addLoad(replica.load());
+    _potentialLeadershipLoadByBrokerId.get(brokerId).addLoad(partition(topicPartition).leader().load());
+  }
+
+  /**
+   * Delete a replica from cluster.
+   * @param topicPartition Topic partition of the replica to be removed.
+   * @param brokerId Id of the broker hosting the replica.
+   */
+  public void deleteReplica(TopicPartition topicPartition, int brokerId) {
+    removeReplica(brokerId, topicPartition);
+    // Update partition info.
+    Partition partition = _partitionsByTopicPartition.get(topicPartition);
+    partition.deleteReplica(brokerId);
+    _replicationFactorByTopic.put(topicPartition.topic(), partition.replicas().size());
+  }
+
+  /**
+   * Refresh the maximal topic replication factor statistic.
+   */
+  public void refreshClusterMaxReplicationFactor() {
```

Comment:
I have considered that way. There is a corner case to consider. 
say we are deleting a replica to make  topic partition's RF from 4 to 3, and the old `_maxReplicationFactor` is 4. In this case, we do not know we should update `_maxReplicationFactor` to 3 or not since we do not know if there are other topic partition still have RF=4. so if we want to do it each time we are deleting a replica, we need to run this heavy global scan.